projects/common/mtl_renderer.cpp

```diff
@@ -0,0 +1,241 @@
+        success = false;
+    }
+
+    if (success) {
+        MTL::CommandQueue* localCommandQueue = pRenderer->Device->newCommandQueue();
+        if (localCommandQueue != nullptr) {
+            pRenderer->Queue = NS::TransferPtr(localCommandQueue);
+        }
+        else {
+            assert(false && "MTL::Device::newCommandQueue() failed");
+            success = false;
+        }
+    }
+
+    pPoolAllocator->release();
+
+    return success;
+}
+
+bool InitSwapchain(
+    MetalRenderer*   pRenderer,
+    void*            cocoaWindow,
+    uint32_t         width,
+    uint32_t         height,
+    uint32_t         bufferCount,
+    MTL::PixelFormat dsvFormat)
+{
+    NS::AutoreleasePool* pPoolAllocator = NS::AutoreleasePool::alloc()->init();
+    bool                 success        = true;
+
+    CGSize layerSize = {(float)width, (float)height};
+
+	// Don't need to Retain/Release the swapchain as the function doesn't match the 
+	// naming templated named in metal-cpp README.md
+    CA::MetalLayer* layer = CA::MetalLayer::layer();
+
+    if (layer != nullptr) {
+        layer->setDevice(pRenderer->Device.get());
+        layer->setPixelFormat(GREX_DEFAULT_RTV_FORMAT);
+        layer->setDrawableSize(layerSize);
+
+        MetalSetNSWindowSwapchain(cocoaWindow, layer);
+
+        pRenderer->Swapchain            = layer;
+        pRenderer->SwapchainBufferCount = bufferCount;
+
+        if (dsvFormat != MTL::PixelFormatInvalid) {
+            for (int dsvIndex = 0; success && (dsvIndex < bufferCount); dsvIndex++) {
+                MTL::TextureDescriptor* pDepthBufferDesc = MTL::TextureDescriptor::alloc()->init();
+
+                pDepthBufferDesc->setPixelFormat(dsvFormat);
+                pDepthBufferDesc->setWidth(width);
+                pDepthBufferDesc->setHeight(height);
+                pDepthBufferDesc->setMipmapLevelCount(1);
+                pDepthBufferDesc->setResourceOptions(MTL::ResourceStorageModePrivate);
+                pDepthBufferDesc->setUsage(MTL::TextureUsageRenderTarget);
+
+                MTL::Texture* localTexture = pRenderer->Device->newTexture(pDepthBufferDesc);
+
+                if (localTexture != nullptr) {
+                    pRenderer->SwapchainDSVBuffers.push_back(NS::TransferPtr(localTexture));
+                }
+                else {
+                    assert(false && "Depth Buffer MTL::Device::newTexture() failed");
+                    success = false;
+                }
+
+                pDepthBufferDesc->release();
+            }
+        }
+    }
+    else {
+        assert(false && "Swapchain creation CA::MetalLayer::layer() failed");
+        success = false;
+    }
+
+    pPoolAllocator->release();
+
+    return success;
+}
+
+NS::Error* CreateBuffer(
+    MetalRenderer* pRenderer,
+    size_t         srcSize,
+    const void*    pSrcData,
+    MetalBuffer*   pBuffer)
+{
+    NS::AutoreleasePool* pPoolAllocator = NS::AutoreleasePool::alloc()->init();
```

Comment:
Any idea why this is needed? It doesn't look like there are any other objects being created besides the buffer.

Reply:
I'm not 100% sure any of the AutoreleasePools are needed? The example code has them so I assume that within their boundaries OSX won't delete any pointers? So the reason I have it is to make sure that OSX doesn't delete the pLocalBuffer before I can transfer it over to the NS::SharedPtr.

But those are all assumptions that may be invalid?